presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1876,60 +1965,116 @@ private Scope analyzeWith(Query node, Scope scope)
             return withScope;
         }
 
-        private void analyzeOrderBy(Query node, Scope scope)
+        private void analyzeOrderBy(Query node, Scope orderByScope)
+        {
+            checkState(node.getOrderBy().isPresent(), "orderBy is absent");
+
+            List<SortItem> sortItems = node.getOrderBy().get().getSortItems();
+            analyzeOrderBy(node, sortItems, orderByScope, Optional.empty(), emptyList());
+        }
+
+        private List<Expression> analyzeOrderBy(QuerySpecification node, Scope orderByScope, List<Expression> outputExpressions)
         {
-            List<SortItem> items = node.getOrderBy()
-                    .map(OrderBy::getSortItems)
-                    .orElse(emptyList());
+            checkState(node.getOrderBy().isPresent(), "orderBy is absent");
 
+            List<SortItem> sortItems = node.getOrderBy().get().getSortItems();
+            return analyzeOrderBy(node, sortItems, orderByScope, Optional.of(node.getSelect()), outputExpressions);
+        }
+
+        private List<Expression> analyzeOrderBy(Node node, List<SortItem> sortItems, Scope orderByScope, Optional<Select> select, List<Expression> outputExpressions)
+        {
             ImmutableList.Builder<Expression> orderByFieldsBuilder = ImmutableList.builder();
 
-            for (SortItem item : items) {
+            for (SortItem item : sortItems) {
                 Expression expression = item.getSortKey();
 
                 if (expression instanceof LongLiteral) {
                     // this is an ordinal in the output tuple
 
                     long ordinal = ((LongLiteral) expression).getValue();
-                    if (ordinal < 1 || ordinal > scope.getRelationType().getVisibleFieldCount()) {
+                    if (ordinal < 1 || ordinal > orderByScope.getRelationType().getVisibleFieldCount()) {
                         throw new SemanticException(INVALID_ORDINAL, expression, "ORDER BY position %s is not in select list", ordinal);
                     }
 
                     expression = new FieldReference(toIntExact(ordinal - 1));
                 }
+                else if (select.isPresent() && select.get().isDistinct()) {
+                    // TODO: use normalized (with respect to qualified names) order by and output expressions
+                    Expression rewrittenOrderByExpression = ExpressionTreeRewriter.rewriteWith(new OrderByExpressionRewriter(extractNamedOutputExpressions(select.get())), expression);
+
+                    int index = outputExpressions.indexOf(rewrittenOrderByExpression);
+                    if (index == -1) {
+                        throw new SemanticException(ORDER_BY_MUST_BE_IN_SELECT, select.get(), "For SELECT DISTINCT, ORDER BY expressions must appear in select list");
+                    }
+
+                    expression = new FieldReference(index);
```

Comment:
We need to preserve original expression here since we need to validate it later with `AggregateAnalyzer`. for instance,
`SELECT a as a, max(a) as c from ... group by max(a)`
in invalid, but here `FieldReference` will be added.